src/main/java/hudson/security/LDAPSecurityRealm.java

```diff
@@ -830,50 +833,64 @@ public LdapUserDetails updateUserDetails(LdapUserDetails d) {
+                GroupDetailsImpl cachedValue = cached.getValue();
+                if (!fetchMembers || cachedValue.getMembers() != null) {
+                    cachedGroup = cachedValue;
+                } else {
+                    cachedGroup = null;
+                }
             } else {
-                cachedGroups = null;
+                cachedGroup = null;
             }
         } else {
-            cachedGroups = null;
+            cachedGroup = null;
         }
 
         // TODO: obtain a DN instead so that we can obtain multiple attributes later
-
-        final Set<String> groups = cachedGroups != null
-                ? cachedGroups
-                : searchForGroupName(groupname);
-        if (cache != null && cachedGroups == null && !groups.isEmpty()) {
+        final GroupDetailsImpl group = cachedGroup != null
+                ? cachedGroup
+                : searchForGroupName(groupname, fetchMembers);
+        if (cache != null && cachedGroup == null) {
             synchronized (this) {
                 if (groupDetailsCache == null) {
-                    groupDetailsCache = new CacheMap<String, Set<String>>(cache.getSize());
+                    groupDetailsCache = new CacheMap<>(cache.getSize());
                 }
-                groupDetailsCache.put(groupname, new CacheEntry<Set<String>>(cache.getTtl(), groups));
+                groupDetailsCache.put(groupname, new CacheEntry<>(cache.getTtl(), group));
             }
         }
 
-        if(groups.isEmpty())
-            throw new UsernameNotFoundException(groupname);
-
-        return new GroupDetailsImpl(fixGroupname(groups.iterator().next()));
+        return group;
     }
 
-    private Set<String> searchForGroupName(String groupname) {
-        Set<String> groups = new TreeSet<>();
+    private @Nonnull GroupDetailsImpl searchForGroupName(String groupname, boolean fetchMembers) throws UsernameNotFoundException, DataAccessException {
         for (LDAPConfiguration conf : configurations) {
             String searchBase = conf.getGroupSearchBase() != null ? conf.getGroupSearchBase() : "";
             String searchFilter = conf.getGroupSearchFilter() != null ? conf.getGroupSearchFilter() : GROUP_SEARCH;
-            groups.addAll(conf.getLdapTemplate().searchForSingleAttributeValues(searchBase, searchFilter, new String[]{groupname}, "cn"));
+            LDAPExtendedTemplate template = conf.getLdapTemplate();
+            GroupDetailsImpl groupDetails = (GroupDetailsImpl)template.searchForFirstEntry(searchBase, searchFilter,
+                    new Object[]{groupname}, new String[]{}, new GroupDetailsMapper());
+            if (groupDetails != null) {
+                if (fetchMembers) {
+                    Set<String> members = conf.getGroupMembershipStrategy().getGroupMembers(groupDetails.getDn(), conf);
+                    groupDetails = new GroupDetailsImpl(groupDetails.getDn(), groupDetails.getName(), members);
```

Comment:
Might need to store the configuration id from where this group came from for future reference. Like what is done with the user details?